### PR TITLE
hipblasTrmm API change to allow inplace and outofplace behaviour

### PR DESCRIPTION
**Pull request by daineAMD**
See [SWDEV-294491](https://ontrack-internal.amd.com/browse/SWDEV-294491).

This PR changes the hipBLAS API to amend the hipblasXtrmm interface. cuBLAS deviates from the legacy BLAS interface by introducing a third matrix to the trmm interface, C. The initial implementation for this in rocBLAS can be seen in https://github.com/ROCmSoftwarePlatform/rocBLAS-internal/pull/850.

We added a deprecation warning for this API in October 2021 in https://github.com/ROCmSoftwarePlatform/hipBLAS/pull/386, this PR realizes this change. We may want to wait until the next major release for this, but I thought I'd open the PR with the changes now. We can discuss next meeting with Braga and all.

### CODE REVIEW

**Review by TorreZuk — APPROVED** [review]
Looks okay if we have deprecated long enough.  Minor questions made

File: library/include/hipblas.h [hunk 1]
```diff
@@ -14438,26 +14438,22 @@ HIPBLAS_EXPORT hipblasStatus_t hipblasZhemmStridedBatched(hipblasHandle_t
                                                           int                         batchCount);
 //! @}
 
-// clang-format off
-HIPBLAS_DEPRECATED_MSG("The hipblasXtrmm API, along with batched versions, will \
-be changing in a future release to allow in-place and out-of-place behavior. This change \
-will introduce an output matrix 'C', matching the rocblas_xtrmm_outofplace API and the \
-cublasXtrmm API.")
-// clang-format on
```
[the remark this answers is not in the document]
Comment (daineAMD) [reply library/include/hipblas.h:14446]: [First seen in ROCm 5.0](https://github.com/ROCmSoftwarePlatform/hipBLAS/blob/release/rocm-rel-5.0/library/include/hipblas.h#L13501).

File: library/include/hipblas.h [hunk 2]
```diff
@@ -14531,15 +14527,20 @@ cublasXtrmm API.")
...
     ldb    [int]
            ldb specifies the first dimension of B. ldb >= max( 1, m ).
 
+    @param[in]
+    C      Device pointer to the matrix C of dimension (ldc, n) on the GPU.
+           Users can pass in the same matrix B to parameter C to achieve
+           in-place functionality of trmm.
+    @param[in]
+    ldc    [int]
+           ldc specifies the first dimension of C. ldc >= max( 1, m ).
```
[the remark this answers is not in the document]
Comment (daineAMD) [reply library/include/hipblas.h:14542]: Hmm, I don't think rocBLAS or cuBLAS does any checking. I'll double check this and add comments, and can change behaviour to be consistent.

[the remark this answers is not in the document]
Comment (amcamd) [reply library/include/hipblas.h:14542]: Interesting question. in-place gives the legacy BLAS functionality where B and C matrices are the same, so I would expect in-place to require ldb==ldc. AFAIK neither the in-place or out-of-place algorithm will not work if C == B but ldb != ldc. cuBLAS does not specify that if B==C then ldb must equal ldc to get the in place functionality. I expect we should return rocblas_status_invalid_size if B==C and ldb != ldc.

[the remark this answers is not in the document]
Comment (daineAMD) [reply library/include/hipblas.h:14542]: cuBLAS doesn't seem to have any checking for ldb == ldc if B == C, i.e. I'm able to run cuBLAS trmm successfully with B == C and ldb != ldc.
rocBLAS doesn't do any checking for this at the moment either. I can add a comment regarding this in this file.
We can discuss if we want to deviate during tomorrow's meeting.

[the remark this answers is not in the document]
Comment (TorreZuk) [reply library/include/hipblas.h:14542]: You can look at rocblas_gemm_ex.hpp validateArgs( )  I was the one who added these checks only 7 months ago.  There was likely a ticket if you want me to find it, but was recent rocblas PR #998.  Would think we could follow this pattern to be consistent.

**Comment by TorreZuk** [comment]
What is going on with this PR?

**Comment by daineAMD** [comment]
Will be replaced by Andrew's PR as I understand.

**Comment by amcamd** [comment]
I forgot about this PR, I am planning to add it like the deprecation in rocBLAS. It will be possible to change behavior with -DHIPBLAS_V1.

**Comment by amcamd** [comment]
Please leave this PR open.

**Comment by daineAMD** [comment]
Closing in favour of #617

**Pull request closed without merging**